Setup/UpgradeData.php

```diff
@@ -4,33 +4,83 @@
 use Magento\Framework\Setup\UpgradeDataInterface;
 use Magento\Framework\Setup\ModuleDataSetupInterface;
 use Magento\Framework\Setup\ModuleContextInterface;
-use Magento\Framework\Filesystem\DirectoryList;
 
 class UpgradeData implements UpgradeDataInterface
 {
     /**
-     * DirectoryList instance
-     * @var \Magento\Framework\Filesystem\DirectoryList $_dir
+     * @var \Magento\Framework\Encryption\EncryptorInterface
      */
-    protected $_dir;
+    protected $_encryptor;
 
+    /**
+     * Logging helper
+     * @var \Klaviyo\Reclaim\Helper\Logger
+     */
+    protected $_klaviyoLogger;
+
+    /**
+     * @var \Klaviyo\Reclaim\Helper\ScopeSetting
+     */
+    protected $_klaviyoScopeSetting;
+
+    /**
+     * @var \Magento\Framework\App\State 
+     */
+    protected $_state;
+
+    /**
+     * @param \Magento\Framework\Encryption\EncryptorInterface $encryptor
+     * @param \Klaviyo\Reclaim\Helper\Logger $klaviyoLogger
+     * @param \Klaviyo\Reclaim\Helper\ScopeSetting $klaviyoScopeSetting
+     * @param \Magento\Framework\App\State $state
+     */
     public function __construct(
-        DirectoryList $dir
+        \Magento\Framework\Encryption\EncryptorInterface $encryptor,
+        \Klaviyo\Reclaim\Helper\Logger $klaviyoLogger,
+        \Klaviyo\Reclaim\Helper\ScopeSetting $klaviyoScopeSetting,
+        \Magento\Framework\App\State $state
     )
     {
-        $this->_dir = $dir;
+        $this->_encryptor = $encryptor;
+        $this->_klaviyoLogger = $klaviyoLogger;
+        $this->_klaviyoScopeSetting = $klaviyoScopeSetting;
+        $this->_state = $state;
     }
 
     public function upgrade(
         ModuleDataSetupInterface $setup,
         ModuleContextInterface $context
     )
     {
+        try{
+            $this->_state->getAreaCode();
+        }
+        catch (\Magento\Framework\Exception\LocalizedException $ex) {
+            $this->_state->setAreaCode(\Magento\Framework\App\Area::AREA_ADMINHTML);
+        }
+
+        $setup->startSetup();
+
         //Klaviyo log file creation
-        $path = $this->_dir->getPath('log') . '/klaviyo.log';
+        $path = $this->_klaviyoLogger->getPath();
         if (!file_exists($path)) {
             fopen($path, 'w');
         }
         chmod($path, 0644);
+
+        /**
+         * in release 1.1.7 we started using the encrypted backend model for the private api key
+         * this check ensures that when upgrading to this version the key is stored properly
+         * otherwise we'd be trying to decrypt an unencrypted value elsewhere in the extension code (yikes)
+         */
+        if (version_compare($context->getVersion(), '1.1.7', '<')) {
```

Comment:
can we make that a const?

Reply:
the version # ? or something else

Reply:
due to the way that PHP's memory management works (it's Not Great) I've been trying to avoid const-ing strings unless I use it more than once